### DISCUSSION TITLE
Make string comparison generically case-insensitive?

### DISCUSSION THREAD

**Discussion by jswelling** [zeroSteiner/rule-engine#77] [Q&A]
Hi folks;
  I have an application where I'd like to modify the behavior of string comparison so that it is always case insensitive.  I know that I can just add .as_lower to every string in every comparison operation, but I'm hoping to avoid the additional verbiage.  Is there a way to modify the resolvers or some other component to accomplish this?

**Comment by zeroSteiner** [comment]
Yeah a resolver would probably work in your case. There will likely be edge cases though like to_upper would still work and any functions that return strings would be unaffected. Going the resolver route would basically let you normalize the input to upper or lower case.

**Reply by jswelling** [reply]
Would this be as simple as initializing the Context with a resolver function that would lowercase any value that was a string?  I'm really unclear on how resolvers work and have been unable to find good examples.

**Reply by zeroSteiner** [reply]
Alright, I looked into this a bit more closely and on second thought I don't think a resolver is going to do it. The issue is with compound types, the resolver isn't used for member values. In the following example you'll see `name` works just fine but members of the `affiliations` array do not.

```
In [1]: import_rule_engine
   ...: def my_resolver(thing, name):
   ...:     value = rule_engine.resolve_item(thing, name)
   ...:     if isinstance(value, str):
   ...:         value = value.lower()
   ...:     return value
   ...: context = rule_engine.Context(resolver=my_resolver)

In [2]: thing = { 'name': 'Luke Skywalker', 'affiliations': ['Rebel Alliance', 'Jedi Order']}

In [3]: rule_engine.Rule('name == "luke skywalker"', context=context).evaluate(thing)
Out[3]: True

In [4]: rule_engine.Rule('affiliations[0] == "rebel alliance"', context=context).evaluate(thing)
Out[4]: False

In [5]: rule_engine.Rule('affiliations[0] == "Rebel Alliance"', context=context).evaluate(thing)
Out[5]: True
```

An alternative solution might be to set the `regex_flags` of the context to `re.IGNORECASE`, then use the regex match operator `=~`. The main catch here is that the right side of the operator is a regex and can't contain invalid regex characters so you wouldn't want to use a variable. If you're searching simple strings though it might do the trick.

```
In [1]: import re, rule_engine

In [2]: context = rule_engine.Context(regex_flags=re.IGNORECASE)

In [3]: thing = { 'name': 'Luke Skywalker', 'affiliations': ['Rebel Alliance', 'Jedi Order']}

In [4]: rule_engine.Rule('affiliations[0] =~ "rebel alliance"', context=context).evaluate(thing)
Out[4]: True

In [5]: rule_engine.Rule('name =~ "luke skywalker"', context=context).evaluate(thing)
Out[5]: True
```

**Reply by jswelling** [reply]
Thank you; the regex suggestion may solve my problem!
